### PR TITLE
TC-262 TC-416 Rework how ReconnectingWebsocket periodically checks health of ws

### PR DESCRIPTION
**Pull request by nickbrowne**
Combined all timeouts and intervals into one so they can't stomp on each others toes and get the instance into a weird state.

Previously we relied too much on the internal WebSocket instance transitioning from `CLOSING` to `CLOSED` in a timely manner. While in Firefox that happens pretty quickly, Chrome seems to leave it in a `CLOSING` state for an indeterminate amount of time (sometimes it's quick, other times it never seems to transition). So the biggest improvement is moving the reconnect logic away from that, and instead basing it either on being stuck in a particular state, or if there is no heartbeat response from the server.

I also added a lot more debugging (when debug is true), and left quite detailed comments explaining the key parts.

Based the "decisions" about what to do on the `readyState`, which all happens in one place, so it's much easier to see what's going on.

Basically:

 - If the socket is stuck connecting, we close
 - If the socket does not get a heartbeat response, we close
 - If the socket is closed, we reconnect

### CODE REVIEW

**Comment by nickbrowne** [comment]
I've run out of time to test this on staging today like I'd planned (got debug turned on), but I'll do it first thing tomorrow.

Of course feel free to look over my changes to see if they make sense and if there's anything I missed!

**Comment by nickbrowne** [comment]
I've given this a good run through on staging for the last few days. 

Various combinations of yanking network cable, switching wifi networks, tethering, etc - in all cases these changes are able to pick up the change in network conditions and reconnect much more quickly than before.

File: src/client/reconnecting_websocket.coffee [hunk 1]
```diff
@@ -144,27 +108,100 @@ class ReconnectingWebSocket
...
+
+  _handleWebsocketClose: (event) =>
+    @ws = null
+    @readyState = WebSocket.CLOSED
+    @onclose event
+
+    if @forcedClose
+      # if we're closing for good, kill the health check interval so it
+      # doesn't try to re-establish the connection
+      clearInterval @healthCheckInterval
```
Comment (tiagoamaro) [inline src/client/reconnecting_websocket.coffee]: "if we're closing for good, kill the health check interval so it" 👏

File: src/client/reconnecting_websocket.coffee [hunk 2]
```diff
@@ -144,27 +108,100 @@ class ReconnectingWebSocket
...
+      when WebSocket.OPEN then @_checkHeartbeat()
+
+  _connect: =>
+    @ws = new WebSocket(@url)
+    @readyState = WebSocket.CONNECTING
+    console.log "ReconnectingWebSocket", "attempt-connect", @url  if @debug
+    @ws.addEventListener "open", @_handleWebsocketOpen
+    @ws.addEventListener "close", @_handleWebsocketClose
+    @ws.addEventListener "message", @_handleWebsocketMessage
+    @ws.addEventListener "error", @_handleWebsocketError
```
Comment (tiagoamaro) [inline src/client/reconnecting_websocket.coffee]: Actually creating a new connection is way better than what was happening when trying to rescue the dropped one! 👏
